include/dsn/utility/nth_element.h

```diff
@@ -0,0 +1,115 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <cstdint>
+#include <functional>
+#include <limits>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <dsn/c/api_utilities.h>
+#include <dsn/dist/fmt_logging.h>
+#include <dsn/utility/ports.h>
+
+namespace dsn {
+
+// The finder helps to find multiple nth elements of a sequence container (e.g., std::vector)
+// at a time, based on nth_element() of STL.
+template <typename T,
+          typename Compare = std::less<T>,
+          typename ContainerType = typename std::vector<T>>
+class stl_nth_element_finder
+{
+public:
+    using value_type = T;
+    using container_type = ContainerType;
+    using size_type = typename container_type::size_type;
+    using nth_container_type = std::vector<size_type>;
```

Comment:
How about use std::set instead of std::vector? then you can avoid to sort indexes, and `set_nths()` and `nths()` are consistency in semantics.